src/serialize.cpp

```diff
@@ -175,42 +175,27 @@ int SerializePart(google::protobuf::Message *message, Local<Object> subj,
   const Reflection *r = message->GetReflection();
   const Descriptor *d = message->GetDescriptor();
 
-  // build a list of required properties
-  vector<string> required;
   for (int i = 0; i < d->field_count(); i++) {
     const FieldDescriptor *field = d->field(i);
-    if (field->is_required())
-      required.push_back(field->name());
-  }
-
-  // build a reflection
-  // get properties of passed object
-  Local<Array> properties = subj->GetPropertyNames();
-  uint32_t len = properties->Length();
-
-  // check that all required properties are present
-  for (uint32_t i = 0; i < required.size(); i++) {
-    Local<String> key =
-        Nan::New<String>(required.at(i).c_str()).ToLocalChecked();
-    if (!subj->Has(key))
-      return -1;
-  }
 
-  for (uint32_t i = 0; i < len; i++) {
-    Local<Value> property = properties->Get(i);
-    Local<String> property_s = property->ToString();
+    Local<String> fieldKey =
+        Nan::New<String>(field->name().c_str()).ToLocalChecked();
 
-    if (*property_s == NULL)
-      continue;
+    Nan::MaybeLocal<Value> maybeVal = Nan::Get(subj, fieldKey);
 
-    String::Utf8Value temp(property);
-    std::string propertyName = std::string(*temp);
+    if (maybeVal.IsEmpty() && field->is_required()) {
+      return -1;
```

Comment:
I think I'm missing a check here...

Should probably be:

```c++
if (maybeVal.IsEmpty()) {
if (field->is_required()) {
return -1;
} else {
continue;
}
}
```

Though in my testing this never resolved to true even if I didn't pass the field. I think the maybe value resolves to undefined, and that check is handled below.